src/Nonlinear/Nonlinear.jl

```diff
@@ -0,0 +1,515 @@
+import NaNMath:
+    sin,
+    cos,
+    tan,
+    asin,
+    acos,
+    acosh,
+    atanh,
+    log,
+    log2,
+    log10,
+    lgamma,
+    log1p,
+    pow,
+    sqrt
+
+include("univariate_expressions.jl")
+include("operators.jl")
+include("types.jl")
+include("parse.jl")
+
+"""
+    set_objective(data::NonlinearData, obj)::Nothing
+
+Parse `obj` into a [`NonlinearExpression`](@ref) and set as the objective
+function of `data`.
+
+`obj` must be a type that is supported by [`parse_expression`](@ref).
+
+## Examples
+
+```julia
+data = NonlinearData()
+x = MOI.VariableIndex(1)
+set_objective(data, :(\$x^2 + 1))
+set_objective(data, x)
+```
+"""
+function set_objective(data::NonlinearData, obj)
+    data.objective = parse_expression(data, obj)
+    return
+end
+
+"""
+    add_expression(data::NonlinearData, expr)::ExpressionIndex
+
+Parse `expr` into a [`NonlinearExpression`](@ref) and add to `data`. Returns an
+[`ExpressionIndex`](@ref) that can be interpolated into other input expressions.
+
+`expr` must be a type that is supported by [`parse_expression`](@ref).
+
+## Examples
+
+```julia
+data = NonlinearData()
+x = MOI.VariableIndex(1)
+ex = add_expression(data, :(\$x^2 + 1))
+set_objective(data, :(sqrt(\$ex)))
+```
+"""
+function add_expression(data::NonlinearData, expr)
+    push!(data.expressions, parse_expression(data, expr))
+    return ExpressionIndex(length(data.expressions))
+end
+
+function Base.getindex(data::NonlinearData, index::ExpressionIndex)
+    return data.expressions[index.value]
+end
+
+"""
+    add_constraint(data::NonlinearData, input::Expr)::ConstraintIndex
+
+Parse `input` into a [`NonlinearConstraint`](@ref) and add to `data`. Returns a
+[`ConstraintIndex`](@ref) that can be used to delete the constraint or query
+solution information.
+
+## Examples
+
+```julia
+data = NonlinearData()
+x = MOI.VariableIndex(1)
+c = add_constraint(data, :(\$x^2 <= 1))
+```
+"""
+function add_constraint(data::NonlinearData, input::Expr)
+    expr, set = _expr_to_constraint(input)
+    f = parse_expression(data, expr)
+    data.last_constraint_index += 1
+    index = ConstraintIndex(data.last_constraint_index)
+    data.constraints[index] = NonlinearConstraint(f, set)
+    return index
+end
+
+"""
+    delete(data::NonlinearData, c::ConstraintIndex)::Nothing
+
+Delete the constraint index `c` from `data`.
+
+## Examples
+
+```julia
+data = NonlinearData()
+x = MOI.VariableIndex(1)
+c = add_constraint(data, :(\$x^2 <= 1))
+delete(data, c)
+```
+"""
+function delete(data::NonlinearData, c::ConstraintIndex)
+    delete!(data.constraints, c)
+    return
+end
+
+"""
+    row(data::NonlinearData, c::ConstraintIndex)::Int
+
+Return the 1-indexed row of the constraint index `c` in `data`.
+
+## Examples
+
+```julia
+data = NonlinearData()
+x = MOI.VariableIndex(1)
+c = add_constraint(data, :(\$x^2 <= 1))
+row(data, c)  # Returns 1
+```
+"""
+function row(data::NonlinearData, c::ConstraintIndex)
```

Comment:
It's used in JuMP when calling `dual(::NonlinearConstraintRef)`, because we need to map from constraint to the 1-indexed row/ordinal index. Caching a map is another option.